src/change_stream/mod.rs

```diff
@@ -227,11 +220,47 @@ where
     T: DeserializeOwned + Unpin + Send + Sync,
 {
     fn poll_next_in_batch(&mut self, cx: &mut Context<'_>) -> Poll<Result<BatchValue>> {
+        if let Some(mut pending) = self.pending_resume.take() {
+            match Pin::new(&mut pending).poll(cx) {
+                Poll::Pending => {
+                    self.pending_resume = Some(pending);
+                    return Poll::Pending;
+                }
+                Poll::Ready(Ok(new_stream)) => {
+                    // Ensure that the old cursor is killed on the server selected for the new one.
+                    self.cursor
+                        .set_drop_address(new_stream.cursor.address().clone());
+                    self.cursor = new_stream.cursor;
+                    self.args = new_stream.args;
+                    return Poll::Pending;
+                }
+                Poll::Ready(Err(e)) => return Poll::Ready(Err(e)),
+            }
+        }
         let out = self.cursor.poll_next_in_batch(cx);
-        if let Poll::Ready(Ok(bv)) = &out {
-            if let Some(token) = get_resume_token(bv, self.cursor.post_batch_resume_token())? {
-                self.resume_token = Some(token);
+        match &out {
+            Poll::Ready(Ok(bv)) => {
+                if let Some(token) = get_resume_token(bv, self.cursor.post_batch_resume_token())? {
+                    self.data.resume_token = Some(token);
+                }
+                if matches!(bv, BatchValue::Some { .. }) {
+                    self.data.document_returned = true;
+                }
+            }
+            Poll::Ready(Err(e)) if e.is_resumable() && !self.data.resume_attempted => {
+                self.data.resume_attempted = true;
+                let client = self.cursor.client().clone();
+                let args = self.args.clone();
+                let data = self.data.clone();
+                self.pending_resume = Some(Box::pin(async move {
+                    let new_stream: Result<ChangeStream<ChangeStreamEvent<()>>> = client
```

Comment:
One tricky part here is the following requirement from the spec:

>If the server supports sessions, the resume attempt MUST use the same session as the previous attempt's command.

So I think this applies to implicit sessions too.

Reply:
Updated to do so.  This could have avoided a new method in `GetMoreProvider` by directly exposing the provider in `GenericCursor` which would allow `Cursor` to call methods on `ImplicitGetMoreProvider`; maintaining the encapsulation seemed worth the extra boilerplate, but I'm uncertain of that tradeoff.